controllers/virtualmachinesetresourcepolicy/v1alpha2/virtualmachinesetresourcepolicy_controller.go

```diff
@@ -0,0 +1,164 @@
+}
+
+// ReconcileNormal reconciles a VirtualMachineSetResourcePolicy.
+func (r *Reconciler) ReconcileNormal(ctx *context.VirtualMachineSetResourcePolicyContextA2) error {
+	if !controllerutil.ContainsFinalizer(ctx.ResourcePolicy, finalizerName) {
+		// Return here so the VirtualMachineSetResourcePolicy can be patched immediately. This ensures that
+		// the resource policies are cleaned up properly when they are deleted.
+		controllerutil.AddFinalizer(ctx.ResourcePolicy, finalizerName)
+		return nil
+	}
+
+	ctx.Logger.Info("Reconciling VirtualMachineSetResourcePolicy")
+	defer func() {
+		ctx.Logger.Info("Finished Reconciling VirtualMachineSetResourcePolicy")
+	}()
+
+	if err := r.VMProvider.CreateOrUpdateVirtualMachineSetResourcePolicy(ctx, ctx.ResourcePolicy); err != nil {
+		ctx.Logger.Error(err, "Provider failed to reconcile VirtualMachineSetResourcePolicy")
+		return err
+	}
+
+	return nil
+}
+
+// deleteResourcePolicy deletes a VirtualMachineSetResourcePolicy resource.
+func (r *Reconciler) deleteResourcePolicy(ctx *context.VirtualMachineSetResourcePolicyContextA2) error {
+	resourcePolicy := ctx.ResourcePolicy
+
+	// Skip deleting a VirtualMachineSetResourcePolicy if it is referenced by a VirtualMachine.
+	vmsInNamespace := &vmopv1.VirtualMachineList{}
+	if err := r.List(ctx, vmsInNamespace, client.InNamespace(resourcePolicy.Namespace)); err != nil {
+		ctx.Logger.Error(err, "Failed to list VMs in namespace", "namespace", resourcePolicy.Namespace)
+		return err
+	}
+
+	for _, vm := range vmsInNamespace.Items {
+		if vm.Spec.Reserved.ResourcePolicyName == resourcePolicy.Name {
+			return fmt.Errorf("failing VirtualMachineSetResourcePolicy deletion since VM: '%s' is referencing it, resourcePolicyName: '%s'",
+				vm.NamespacedName(), resourcePolicy.NamespacedName())
+		}
+	}
+
+	ctx.Logger.V(4).Info("Attempting to delete VirtualMachineSetResourcePolicy")
+	if err := r.VMProvider.DeleteVirtualMachineSetResourcePolicy(ctx, resourcePolicy); err != nil {
+		ctx.Logger.Error(err, "error in deleting VirtualMachineSetResourcePolicy")
+		return err
+	}
+	ctx.Logger.Info("Deleted VirtualMachineSetResourcePolicy successfully")
+
+	return nil
+}
+
+func (r *Reconciler) ReconcileDelete(ctx *context.VirtualMachineSetResourcePolicyContextA2) error {
+	ctx.Logger.Info("Reconciling VirtualMachineSetResourcePolicy Deletion")
+	defer func() {
+		ctx.Logger.Info("Finished Reconciling VirtualMachineSetResourcePolicy Deletion")
+	}()
+
+	if controllerutil.ContainsFinalizer(ctx.ResourcePolicy, finalizerName) {
+		if err := r.deleteResourcePolicy(ctx); err != nil {
+			return err
+		}
+
+		controllerutil.RemoveFinalizer(ctx.ResourcePolicy, finalizerName)
+	}
+
+	return nil
+}
+
+// -kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinesetresourcepolicies,verbs=get;list;watch;create;update;patch;delete
+// -kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinesetresourcepolicies/status,verbs=get;update;patch
```

Comment:
Any difference between `// -kubebuilder` and `// +kubebuilder`? I couldn't find any reference to the minus tag in the kubebuilder doc...

Reply:
It was just so that controller-gen would not process the directive. For the controllers though that isn't needed since it is the same for v1a1 and v1a2. I had updated the other controllers but had just missed this one.

The `-kubebuilder` is used in the webhooks since we want to disable those directives until we get around to generating separate v1a1 and v1a2 manifests.

Reply:
I see, that makes sense.